liquidation-monitor/src/autotask.js

```diff
@@ -0,0 +1,111 @@
+const axios = require('axios');
+
+async function post(url, method, headers, data) {
+  return axios({
+    url,
+    method,
+    headers,
+    data,
+  });
+}
+
+async function postToDiscord(url, message) {
+  const method = 'post';
+  const headers = {
+    'Content-Type': 'application/json',
+  };
+  const data = JSON.stringify({ content: message });
+
+  let response;
+  try {
+    // perform the POST request
+    response = await post(url, method, headers, data);
+  } catch (error) {
+    // is this a "too many requests" error (HTTP status 429)
+    if (error.response && error.response.status === 429) {
+      // the request was made and a response was received
+      // try again after waiting 5 seconds
+      // eslint-disable-next-line no-promise-executor-return
+      const promise = new Promise((resolve) => setTimeout(resolve, 5000));
+      await promise;
+      response = await post(url, method, headers, data);
+    } else {
+      // re-throw the error if it's not from a 429 status
+      throw error;
+    }
+  }
+  return response;
+}
+
+// eslint-disable-next-line func-names
+exports.handler = async function (autotaskEvent) {
+  // ensure that the autotaskEvent Object exists
+  if (autotaskEvent === undefined) {
+    return {};
+  }
+
+  const { secrets } = autotaskEvent;
+  if (secrets === undefined) {
+    return {};
+  }
+
+  // ensure that there is a DiscordUrl secret
+  const { discordUrl } = secrets;
+  if (discordUrl === undefined) {
+    return {};
+  }
+
+  // ensure that the request key exists within the autotaskEvent Object
+  const { request } = autotaskEvent;
+  if (request === undefined) {
+    return {};
+  }
+
+  // ensure that the body key exists within the request Object
+  const { body } = request;
+  if (body === undefined) {
+    return {};
+  }
+
+  // ensure that the alert key exists within the body Object
+  const { alert } = body;
+  if (alert === undefined) {
+    return {};
+  }
+
+  // ensure that the alert key exists within the body Object
+  const { source } = body;
+  if (source === undefined) {
+    return {};
+  }
```

Comment:
This was an initial approach we took to avoid flooding people's email addresses with Autotask failures (i.e. to avoid alert fatigue), but thinking back on it, it probably is not the correct approach.  Each of these will silently swallow any potential issues, which seems worse (apologies for leaving the legacy code in place for so long).

Updating each of these conditionals to throw with some indication of the issue is what the new(er) Autotasks are doing.  This example is for an Autotask triggered by a Contract Sentinel:
```javascript
  if (autotaskEvent === undefined) {
    throw new Error('autotaskEvent undefined');
  }

  const { secrets, request } = autotaskEvent;
  if (secrets === undefined) {
    throw new Error('secrets undefined');
  }

  // ensure that there is a DiscordUrl secret
  const { GovernanceDiscordUrl: discordUrl } = secrets;
  if (discordUrl === undefined) {
    throw new Error('GovernanceDiscordUrl undefined');
  }

  // ensure that the request key exists within the autotaskEvent Object
  if (request === undefined) {
    throw new Error('request undefined');
  }

  // ensure that the body key exists within the request Object
  const { body } = request;
  if (body === undefined) {
    throw new Error('body undefined');
  }

  // ensure that the alert key exists within the body Object
  const {
    matchReasons,
    hash: transactionHash,
  } = body;
  if (matchReasons === undefined) {
    throw new Error('matchReasons undefined');
  }
```

In this way we are at least controlling when the error is thrown and what information it is provided.  If we didn't do it this way, we might make it further into the logic and then end up with some other error that may be harder to trace.  I'm also open to suggestions on other approaches besides all of this error throwing.

Reply:
Fixed in latest revision